File: Python/Programmers/Level 2/올바른 괄호.py
# 열린 괄호로 시작해 닫힌 괄호를 만나면 pop 즉, stack 사용

def solution(s):
    stack = []
    for b in s:
        if stack and stack[-1] == "(" and b == ")": # stack이 비어있지 않거나 stack[-1]이 열린 괄호이고, b가 닫힌 괄호이면 올바른 괄호이므로
            stack.pop() # pop
            continue
        
        stack.append(b) # 그렇지 않으면 append
    
    return len(stack) == 0

File: JavaScript/Programmers/Level 2/올바른 괄호.js
// "(", ")"의 개수가 똑같은지 확인하면 됨, stack 사용

function solution(s){
    const answer = [];
    for (let b of s) { // s 문자열 모든 요소 순회
        if (answer.length && answer[answer.length - 1] === "(" && b === ")") { // answer이 비어있지 않고, answer의 마지막 요소가 "(", b가 ")"면 짝이 맞는 것이므로
            answer.pop(); // pop 연산
            continue; // continue
        }
        
        if (b === ")") { // 괄호 시작이 ")"로 시작한다면 false return
            return false;
        }
        
        answer.push(b); // "(" push
    }
    
    return answer.length ? false : true; // answer가 비어있지 않으면 false, 비어있으면 true return
}
